### ISSUE TITLE
What exactly are the precision and recall reported by yolov8, for a multiclass object detection task? Is it possible to get them per class?

### ISSUE THREAD

**Issue by AndreaPi** [ultralytics/ultralytics#3662]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm happily training various yolov8 models with your great library, on the task of **multiclass object detection.** Among the various plots I see in Tensorboard, I would like to know more about precision and recall:

![image](https://github.com/ultralytics/ultralytics/assets/8233615/530502ec-c463-4586-9e1b-4ae672184ee0)
![image](https://github.com/ultralytics/ultralytics/assets/8233615/b660ddc9-e44f-4dd4-a14a-6f07eda4b4fd)

In a single class object detection setting, the meaning of precision and recall is pretty obvious. However, in multiclass settings, it becomes ambiguous, and a choice must be made. 

1. which kind of precision/recall are you using? Are you using macro-averaged precision and recall, or micro-averaged? https://towardsdatascience.com/micro-macro-weighted-averages-of-f1-score-clearly-explained-b603420b292f It should be macro-averaged, since micro-averaged precision and recall are the same (see link), but the P & R you compute are different, so they can't be micro (unless there's a bug in the code!)
2. is it possible to get precision and recall **for each class**? This would be extremely useful to compare `ultralytics` vs other frameworks I'm testing. I'm looking for precision and recall **at a specified confidence threshold** (not the average precision/recall for each class, computed over an array of confidence thresholds). But getting at least AP and AR per class would be already a good start.

PS I had a look at the documentation, and there's nothing about micro/macro averaging.

![image](https://github.com/ultralytics/ultralytics/assets/8233615/91578465-9071-4678-bef4-d3a8809be428)

**Comment by bit-scientist** [comment]
you can see your validation set being evaluated on the `best.pt` model after training similar to below:
```
Model summary (fused): 168 layers, 11126358 parameters, 0 gradients
  Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 121/121 [00:48<00:00,  2.49it/s]
    all       7684      13383      0.803      0.776      0.823      0.596
   bike       7684       9215      0.833      0.752      0.827      0.601
vehicle       7684       4168      0.774      0.801      0.818      0.592

```
Alternatively, `yolo detect val model=path/to/best.pt` should also output those metrics per class.

**Comment by glenn-jocher** [comment]
Hi @AndreaPi, thank you for your question. In YOLOv8, the validation set can be evaluated on the `best.pt` model after training. The metrics are printed to the screen and can also be retrieved from file.

To get the precision and recall per class, you can use the `yolo detect val model=path/to/best.pt` command. This command will output the metrics, including precision and recall, for each class in the validation set.

I understand that you need to retrieve the metrics from file, as you launch multiple trainings as batch jobs on a headless server. I encourage you to try the `yolo detect` command in the `val` mode and check if it provides the desired precision and recall metrics per class. If it does, you can close the issue.

If you have any further questions or need additional assistance, please let me know. Happy training!

**Comment by AndreaPi** [comment]
Hi @glenn-jocher,

I'm using this command to run validation:

```
yolo detect val \
split="test" \
model="../../logs/yolov8_large_bs8_300e_imgsz1280_lr1e-3/weights/best.pt" \
conf="0.3" \
data="configs/mydataset.yaml" \
save_json="True" \
project="../../logs" \
name="yolov8_large_bs8_300e_imgsz1280_lr1e-3/test"
```

However, the folder `test` which is created doesn't contain a file with the precision and recall per class:

```
root@fd1e85d54a98:/workspace/.../test# ls
F1_curve.png  PR_curve.png  P_curve.png  R_curve.png  confusion_matrix.png  confusion_matrix_normalized.png  predictions.json  val_batch0_labels.jpg  val_batch0_pred.jpg
```

The `predictions.json` contains the model predictions (on the training set, I guess) but not the precision and recall for class. Is there some way to write them to disk?

**Comment by OgunSerifOnargan** [comment]
Is it possible to display them at the end of each epoch?

**Comment by glenn-jocher** [comment]
@OgunSerifOnargan hello! Thank you for reaching out and for the positive feedback on using YOLOv8. 😊

To answer your questions:

1. The precision and recall metrics reported by YOLOv8 during training are calculated per class and then averaged across all classes, which is a form of macro-averaging. This means each class contributes equally to the final average precision and recall, regardless of the class's frequency in the dataset.

2. Yes, it is possible to get precision and recall for each class individually. When you run the `val` mode, YOLOv8 provides detailed metrics per class, including precision and recall at the specified confidence threshold. You can access these metrics in the results file that is saved after validation (`results.csv` for example), which contains class-wise metrics.

For average precision (AP) and average recall (AR) per class, these are also available in the `results.csv` file after running validation. They are computed over an array of confidence thresholds, providing a comprehensive view of the model's performance across different confidence levels.

If you need further assistance or have more questions, feel free to open an issue in the [Ultralytics YOLOv8 GitHub repository](https://github.com/ultralytics/ultralytics/issues), and we'll be happy to help! 🚀
